[MXNET-1255] update hybridize documentation

roywei opened the pull request:
## Description ##
This is the first step to address issues in:
https://github.com/apache/incubator-mxnet/issues/10875
https://github.com/apache/incubator-mxnet/issues/12100 
https://github.com/apache/incubator-mxnet/issues/12109 

First provide documentation and current work around.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
Add documentations for operators does not work with hybridization

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

piyushghai commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
 
+## Operators that does not work with hybridize
+
+While most APIs are the same in NDArray and Symbol, there are some differences. Writting `F.operator` and call `hybridize` may not work all the time.
```
Nit : ```Writing``` instead of ```writting```

piyushghai commented on docs/tutorials/gluon/hybrid.md:
````diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+```python
+def hybrid_forward(self, F, x):
+    # avoid writting: return x + F.ones((1, 1))
````
Same nit above :)

roywei commented:
@mxnet-label-bot add[Doc, pr-awaiting-review]

ChaiBapchya commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+|[*broadcast_lesser*](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.broadcast_lesser) |	Returns the result of element-wise *lesser than* (<) comparison operation with broadcasting. |
+|[*broadcast_lesser_equal*](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.broadcast_lesser_equal) | Returns the result of element-wise *lesser than or equal to* (<=) comparison operation with broadcasting. |
+
+For example, if you wan to add a NDarray to your input x, use `broadcast_add` instead of `+`:
```
nit: "want"

ChaiBapchya commented on docs/tutorials/gluon/hybrid.md:
````diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+    return x[0]
+```
+
+The current workaround is explicitly call [`slice`](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.slice) operators in `hybrid_forwar`d.
````
nit : `hybrid_forward`

ChaiBapchya replied:
nit: _to_ explicitly call

ChaiBapchya commented on docs/tutorials/gluon/hybrid.md:
````diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+```
+
+### Slice
+[] in NDArray is to get a slice from the array. [] in Symbol is to get an output from a grouped symbol.
````
nit : how about 
`[]` in NDArray is used to get a slice from the array. However, `[]` in Symbol is used to get an output from a grouped symbol.

ChaiBapchya commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+#### In-Place Arithmetic Operators
+
+In place arithmetic operators are also used a lot in Gluon imperative mode. You need to avoid that and write the operations explicitly in `hybrid_forward`.
```
nit : In-place

ChaiBapchya commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+#### In-Place Arithmetic Operators
+
+In place arithmetic operators are also used a lot in Gluon imperative mode. You need to avoid that and write the operations explicitly in `hybrid_forward`.
+For example, avoid `x += y` and use `x  = x + y`, other wise you will get `NotImplementedError`.
```
nit: otherwise

roywei commented:
@aaronmarkham @zheng-da could you help review? Thanks!

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
````diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
 net2 = gluon.SymbolBlock.imports('model-symbol.json', ['data'], 'model-0001.params')
 ```
 
+## Operators that does not work with hybridize
````
```suggestion
## Operators that do not work with hybridize
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
 
+## Operators that does not work with hybridize
+
+If you want to hybridize your model, you must use `F.some_operator` in your 'hybrid_forward' function, F will be 'mxnet.nd' before you hybridize and 'mxnet.sym' after hybridize. While most APIs are the same in NDArray and Symbol,  there are some differences. Writing `F.some_operator` and call `hybridize` may not work all the time.
```
```suggestion
If you want to hybridize your model, you must use `F.some_operator` in your 'hybrid_forward' function. 
`F` will be 'mxnet.nd' before you hybridize and 'mxnet.sym' after hybridize. While most APIs are the same in NDArray and Symbol,  there are some differences. Writing `F.some_operator` and call `hybridize` may not work all of the time.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+### Element-wise Operators
+
+The following arithmetic and comparison APIs are automatically broadcasted if the input NDArrays have different shapes.
+However, that's not the case in Symbol API, it's not automatically broadcasted and you have to manually specify whether to use element-wise operator or broadcast operators.
```
```suggestion
However, that's not the case in Symbol API. It's not automatically broadcasted, and you have to manually specify whether to use element-wise operator or broadcast operators.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+However, that's not the case in Symbol API, it's not automatically broadcasted and you have to manually specify whether to use element-wise operator or broadcast operators.
+
+
+| NDArray APIs  | Description  |
```
This table is a little hard to read. Could the `<=>` be replaced with columns instead?

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+| NDArray APIs  | Description  |
+|---|---|
+| [*NDArray.\__add\__*](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.__add__) | x.\__add\__(y) <=> x+y <=> mx.nd.add(x, y)  |
```
Just a suggestion: You could try using math formatting for better presentation?

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+### Shape
+
+Gluon's imperative interface is very flexible and allows you to print shape of the NDArray. However, Symbol does not have shape attributes. As a result, you need to avoid printing shapes in `hybrid_forward`.
```
```suggestion
Gluon's imperative interface is very flexible and allows you to print the shape of the NDArray. However, Symbol does not have shape attributes. As a result, you need to avoid printing shapes in `hybrid_forward`.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+### Slice
+`[]` in NDArray is used to get a slice from the array. However, `[]` in Symbol is used to get an output from a grouped symbol.
+For example, you will get different result for the following method before and after hybridization.
```
```suggestion
For example, you will get different results for the following method before and after hybridization.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+### Not implemented operators
+
+Some of the often used operators in NDArray  are not implemented in Symbol, and will cause hybridization failure
```
```suggestion
Some of the often used operators in NDArray are not implemented in Symbol, and will cause hybridization failure.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+#### Array API
+
+`mx.nd.array()` is used a lot but Symbol does not have the array API. The current workaround is to use F.ones/ F.zeros/ F.full which exists in both NDArrays and Symbols.
```
```suggestion
`mx.nd.array()` is used a lot, but Symbol does not have the `array` API. The current workaround is to use `F.ones`, `F.zeros`, or `F.full`, which exist in both the NDArray and Symbol APIs.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+#### In-Place Arithmetic Operators
+
+In-place arithmetic operators are also used a lot in Gluon imperative mode. You need to avoid that and write the operations explicitly in `hybrid_forward`.
```
"avoid that"? What exactly? All in-place operators? Can you clarify this?
Maybe better to say something like...
```suggestion
In-place arithmetic operators may be used in Gluon imperative mode, however if you expect to hybridize, you should write the operations explicitly instead.
```

roywei replied:
avoid all in-place operators. I have added example: avoid `x+=y` and use `x = x + y`.

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +137,94 @@ to gluon with `SymbolBlock`:
+
+| NDArray in-place arithmetic operators | Description |
+|---|---|
+|[*NDArray.\__iadd\__*](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.__iadd__) |	x.__iadd__(y) <=> x+=y |
```
Notice that the double underscore is bolding the text. I'd be worried what this will be interpreted as by Sphinx/rst down the line.
You can escape it, or like I mentioned before use `:math:`.

roywei commented:
@aaronmarkham Thanks for the review, I have addressed your comments.
1. I have also added [dive into deeplearning book ](http://en.diveintodeeplearning.org/chapter_computational-performance/hybridize.html) as related content, let me know if it's not ready.
2. For the table format, I was following the format on [website](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.__add__). As for the math formulas,  it seems only renders on website, but no on downloaded jupyter notebooks as this tutorial is donwloadable. Let me knwo if you have other suggestions.
3. Added `asnumpy` and a summary section
Thanks!

roywei commented:
@ThomasDelteil @nswamy Could you help take a look? Thanks!

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -1,6 +1,9 @@
 # Hybrid - Faster training and easy deployment
 
-*Note: a newer version is available [here](http://gluon.mxnet.io/chapter07_distributed-learning/hybridize.html).*
+*Related Contents:*
```
nit: Content instead of Contents (but don't restart CI just for that change).

roywei replied:
I added a link to the [new dive into deeplearning book](http://en.diveintodeeplearning.org/chapter_computational-performance/hybridize.html) so made it "contents"

aaronmarkham replied:
It's a semantic thing I guess. There's probably a rule here but... usually you can say Content is plural already (as it refers to a mass of stuff) and you don't usually say "Related Contents". It's more common to say "Related Content" even if there's more than one item in the list.
It is weird because you do say "Table of Contents", not "Table of Content". 
🤷‍♂️ English is weird. It's not a showstopper... it's fine whichever way you want to do it. (but do it my way). Jk

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +140,106 @@ to gluon with `SymbolBlock`:
+
+### Not implemented operators
+
+Some of the often used operators in NDArray are not implemented in Symbol, and will cause hybridization failure
```
```suggestion
Some of the often used operators in NDArray are not implemented in Symbol, and will cause hybridization failure.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -137,4 +140,106 @@ to gluon with `SymbolBlock`:
+Some of the often used operators in NDArray are not implemented in Symbol, and will cause hybridization failure
+
+#### NDArray.asnumpy
+Symbol does not support `asnumpy` function, you need to avoid calling `asnumpy` in `hybrid_forward`.
```
```suggestion
Symbol does not support the `asnumpy` function. You need to avoid calling `asnumpy` in `hybrid_forward`.
```

aaronmarkham commented on docs/tutorials/gluon/hybrid.md:
```diff
@@ -1,6 +1,9 @@
 # Hybrid - Faster training and easy deployment
 
-*Note: a newer version is available [here](http://gluon.mxnet.io/chapter07_distributed-learning/hybridize.html).*
+*Related Contents:*
```
```suggestion
*Related Content:*
```

Roshrini commented:
Seems like call the comments are addressed.
@sandeep-krishnamurthy Can you merge this PR?
@mxnet-label-bot Update [pr-awaiting-merge]
